cluster.js

```diff
@@ -0,0 +1,62 @@
+'use strict';
+
+const process = require('process');
+const cluster = require('cluster');
+const { config } = require('./constants.js');
+const { watch } = require('fs');
+const logger = require('./loggerFactory.js');
+const Proxy = require('./Proxy.js');
+const ProcessHandler = require('./ProcessHandler.js');
+
+const WORKER_CNT = config.workerVal;
+const activeWorkers = [];
+
+function disconnectOldWorkers() {
+	const len = activeWorkers.length;
+	for (let i = 0; i < len; ++i) {
+		const oldWorker = activeWorkers.shift();
+		oldWorker.disconnect();
+		setTimeout(() => oldWorker.kill(), 60 * 10 * 1000);
+	}
+}
+
+function spawnNewWorkers() {
+	if (cluster.isMaster) {
+		for (let i = 0; i < WORKER_CNT; ++i) {
+			const worker = cluster.fork();
+			worker.on('error', (err) => {
+				logger.error(`Received error event on ${worker.id} : ${err}`);
+			});
+			logger.info(`Created worker with id ${worker.id}`);
+			activeWorkers.push(worker);
+		}
+	}
+}
+
+let currTime = Date.now();
+watch('tmp/restart.txt', () => {
+	if (Date.now() > currTime) {
+		currTime = Date.now();
+		disconnectOldWorkers();
+		spawnNewWorkers();
+	}
```

Comment:
when done `touch tmp/restart` locally, this piece of code  is executed many times.
```
{"level":30,"wasRaw":true,"time":1628235309666,"tags":["info"],"data":"restarted"}                                                             
{"level":30,"wasRaw":true,"time":1628235309667,"tags":["info"],"data":"restarted"}
{"level":30,"wasRaw":true,"time":1628235309667,"tags":["info"],"data":"restarted"}
{"level":30,"wasRaw":true,"time":1628235309670,"tags":["info"],"data":"restarted"}
```
please ensure it happens exactly once always

Reply:
also, somehow number of processes is increasing with every `touch tmp/restart`.
```
ps aux | grep node | grep ws-reconnect-proxy | wc -l
4
7
10
13
16
```

Reply:
yup this is expected since the workers get disconnected & will get killed post timer is completed -> https://github.com/Archish27/ws-reconnect-proxy/blob/878a93c9af09c62975d4ca487cc5297e7c8a0d71/cluster.js#L18

So, every time when you do touch `tmp/restart.txt` the old workers get disconnect & those old workers get killed post the timer is completed

Reply:
let's brainstorm on how we can gracefully restart.
deploys/restarts should not break existing WS connections.
this might be a noob thought, but can we somehow migrate the old connections to new processes?
if not, then we can kill the old processes once every old WS connection is closed. however, this can take up extra resources if there are multiple restarts in a short span of time.

Reply:
also please check how restarts are being handled in pusher

Reply:
So, as discussed, whenever if we need to deploy on proxy for now we will mark that region down & deploy proxy. Just as we do for nginx. Once the deploy is done mark the region as up. We handling region down cases for CDP as well for more details -> https://github.com/browserstack/infra-config/pull/8743

Reply:
Also, we can deploy cdp proxy directly, all the active connections will be pointing to old workers & newer connections will be going new workers

Reply:
Bug fixed for spawning multiple workers